gflownet/envs/crystals/crystal.py

```diff
@@ -0,0 +1,488 @@
+import json
+from enum import Enum
+from typing import Dict, List, Optional, Tuple, Union
+
+import torch
+from torch import Tensor
+from torchtyping import TensorType
+
+from gflownet.envs.base import GFlowNetEnv
+from gflownet.envs.crystals.composition import Composition
+from gflownet.envs.crystals.lattice_parameters import LatticeParameters
+from gflownet.envs.crystals.spacegroup import SpaceGroup
+from gflownet.utils.crystals.constants import TRICLINIC
+
+
+class Stage(Enum):
+    """
+    In addition to encoding current stage, values of this enum are used for padding individual
+    component environment's actions (to ensure they have the same length for tensorization).
+    """
+
+    COMPOSITION = -2
+    SPACE_GROUP = -3
+    LATTICE_PARAMETERS = -4
+
+
+class Crystal(GFlowNetEnv):
+    """
+    A combination of Composition, SpaceGroup and LatticeParameters into a single environment.
+    Works sequentially, by first filling in the Composition, then SpaceGroup, and finally
+    LatticeParameters.
+    """
+
+    def __init__(
+        self,
+        composition_kwargs: Optional[Dict] = None,
+        space_group_kwargs: Optional[Dict] = None,
+        lattice_parameters_kwargs: Optional[Dict] = None,
+        **kwargs,
+    ):
+        self.composition_kwargs = composition_kwargs or {}
+        self.space_group_kwargs = space_group_kwargs or {}
+        self.lattice_parameters_kwargs = lattice_parameters_kwargs or {}
+
+        self.composition = Composition(**self.composition_kwargs)
+        self.space_group = SpaceGroup(**self.space_group_kwargs)
+        # We initialize lattice parameters here with triclinic lattice system to access
+        # all the methods of that environment, but it will have to be reinitialized using
+        # proper lattice system from space group once that is determined.
+        # Triclinic was used because it doesn't force any initial starting angles.
+        self.lattice_parameters = LatticeParameters(
+            lattice_system=TRICLINIC, **self.lattice_parameters_kwargs
+        )
+
+        # 0-th element of state encodes current stage: 0 for composition,
+        # 1 for space group, 2 for lattice parameters
+        self.source = (
+            [0]
+            + self.composition.source
+            + self.space_group.source
+            + self.lattice_parameters.source
+        )
```

Comment:
states (and source) are lists in all other environments. Is it here a tuple for a particular reason?

Reply:
It is a list :) parenthesis are just to group the summation operation.